opendevin/condenser/condenser.py

```diff
@@ -0,0 +1,147 @@
+from opendevin.core.exceptions import (
+    SummarizeError,
+)
+from opendevin.core.logger import opendevin_logger as logger
+from opendevin.events.action import (
+    AgentSummarizeAction,
+)
+from opendevin.llm.messages import Message
+
+from .prompts import (
+    MESSAGE_SUMMARY_WARNING_FRAC,
+    SUMMARY_PROMPT_SYSTEM,
+    parse_summary_response,
+)
+
+
+class CondenserMixin:
+    """Condenses a group of condensable messages as done by MemGPT."""
+
+    def condense(
+        self,
+        messages: list[Message],
+    ):
+        # Start past the system message, and example messages.,
+        # and collect messages for summarization until we reach the desired truncation token fraction (eg 50%)
+        # Do not allow truncation  for in-context examples of function calling
+        token_counts = [
+            self.get_token_count([message])  # type: ignore
+            for message in messages
+            if message.condensable
+        ]
+        message_buffer_token_count = sum(token_counts)  # no system and example message
+
+        desired_token_count_to_summarize = int(
+            message_buffer_token_count * self.config.message_summary_trunc_tokens_frac  # type: ignore
+        )
+
+        candidate_messages_to_summarize = []
+        tokens_so_far = 0
+        for message in messages:
+            if message.condensable:
+                candidate_messages_to_summarize.append(message)
+                tokens_so_far += self.get_token_count([message])  # type: ignore
+            if tokens_so_far > desired_token_count_to_summarize:
+                last_summarized_event_id = message.event_id
+                break
+
+        # TODO: Add functionality for preserving last N messages
+        # MESSAGE_SUMMARY_TRUNC_KEEP_N_LAST = 3
```

Comment:
Just to note, a minor thought, 2 or 4 messages would seem a better default, because a lot of the time, they will be `action, obs, action, obs...`. Not all the time, but well, if we will pick something we might want to target a common behavior.

Reply:
Hmm, makes sense to keep it 2 or 4.